README.md
# Colonial Charter **beta*

After fixing that dent you got after crashlanding on this planet. You finally realise what the mission was: *establish a colonial outpost and exploit its resources for the benefit of your people.*

> Ok so I'm kind of ready to release this as I have it, but it needs tweaking and probably has loads of bugs. If you find anything or want to contribute post/message here or on the forum.

## Production cycle

So the mod adds colonists using the existing production system, they are essentially an item, the only difference being they are not used up, but rather returned as tired after a shift and will need to rest before returning -- a circular production cycle. They plug into the endgame upgrade system by producing space science-packs, the satellite science reward has been nerfed accordingly to encourage building up a large population. 

There is also an infrastructure required to produce the: food, consumer goods and medical supplies that are needed to keep your colonists happily working. 

### Getting the first colonists 

A colony ship has been added that needs to be sent on the rocket, the colonists will appear as a reward for the rocket launch. 

### Types of buildings

**Housing** - Used to change the state of a colonist, from tired to normal, injured to normal, deceased to an empty pod. Requires food/consumer goods and medical supplies. Also used as a place for colonists to reproduce. 

**Workshop** - A place where colonists produce items, technical doc, legal doc, and space science-packs, returned tired with a chance of getting injured.

**Waiting-hall** - Essentially a chest for colonists, 2x2 entity, you can daisy-chain them with smart inserters and circuits to get similar behaviour to a belt. 

## Future ideas/intentions and things to contribute

If anyone wants to add something send it over and I'll add it in, I can do icons/graphics if you guys can figure out the code, I'm pretty bad at that, though the same goes for the graphics if you have an idea for better graphics, the workshop, housing or anything I'll gladly add it in, not the best at photoshop or 3D modelling. 

> If you guys have ideas for balancing the material costs - please post, looking for feedback on this. 

1. Add achievements, 1st colonist, 1000th colonists, 1st native born, that kind of stuff.  
2. Expand tech tree to use legal documents for upgrades to colonists, faster working/more output, etc, flavoured as amendments to the charter. 
3. A better system for colonists death and injuries, currently the output is a 2% chance of death and 98% of a normal colonist, but obviously, these two don't always fire at the same time. In the long run, it levels out but it can be weird to just have a colonist disappear. 
4. Limit the waiting hall to just colonists, couldn't figure out the method to filter a container to just one item group.
5. Graphic for an inserter that looks like a one-way tunnel that only moves colonists, based on the loader entity. I can probably figure out a graphic but no idea how to do the code. 
6. Different sized building, for workshop/housing, make colonies look more interesting. 
7. Make putting colonists in anything but a waiting hall impossible, not sure it's possible. A workaround could be to give them a stack size of 1 so it's impractical, and larger stack size in a waiting hall, like how an artillery wagon has 100 stacks for shells and chest has 1. Again no idea how to do this. 
8. Make it possible to land colonists sooner than a rocket, not sure how but I get that it kind of sucks you need a late game map to even try the mod. 
9. Add more intermediate items the colonists consume, not sure what the right balance would be without making it tedious.
10. Want to limit colonists for they can only rest when they are a tile that is below a pollution threshold. Encourage using trains to transport colonist and have them living far away from the ugly and polluted factory. 
11. Thought about adding a landing location as an entity, so a rocket launch would give landing coordinates that need to be fed into the landing pad which would then produce colonists. Not sure how to approach the animation for something like that. Could even have the lander returned so it can get refurbished and set back. Lots of possibilities here.
12. Adding O2 or vital liquids as input to workshops and housing, not really comfortable enough with the api to add something like this, but I think it could be worthwhile figuring out. 
13. Add consequences for black-outs, possibly something like ruining production, producing injured or deceased colonist, but not sure if it's desirable or possible to do. 
12. ... Something more? 

## Mod compatibility

It should be fully compatible with any existing games, and I intend to make updates compatible with ongoing saves as I'm using it in my play-though as well. 

The only base game files that I modify is the satellite, to change the science reward to encourage building up a large population. There should be nothing stopping the mod from working with many of the others out there.

Why not check out my other [mod](https://mods.factorio.com/mod/FastDischargeAccu), that adds a fast charging/discharging accumulator, useful for setting priorities in a power network. 


## Shout-out to helpful mods

Some mods whose source files were instructive in helping me put this together. 
[Expanded rocket payloads](https://mods.factorio.com/mod/expanded-rocket-payloads)
[Homeworld redux](https://mods.factorio.com/mod/homeworld_redux)

Finally why not check out my other [url=https://mods.factorio.com/mod/FastDischargeAccu]mod[/url], that adds a fast charging/discharging accumulator, useful for setting priorities in a power network. 

changelog.txt
29-11-2018 - 0.1.1  Fixing legal doc science labs bug. 
18-11-2018 - 0.0.1  Initial version. 

src/prototypes/entity/entities.lua
data:extend({
  -- COLONIAL HOUSING 
    {
    type = "assembling-machine",
    name = "colonial-housing",
    icon = "__ColonialCharter__/graphics/icons/colonial-housing.png",
    icon_size = 32,
    flags = {"placeable-neutral", "placeable-player", "player-creation"},
    minable = {hardness = 0.2, mining_time = 0.5, result = "colonial-housing"},
    max_health = 300,
    corpse = "big-remnants",
    dying_explosion = "medium-explosion",
    collision_box = {{-1.2, -1.2}, {1.2, 1.2}},
    selection_box = {{-1.5, -1.5}, {1.5, 1.5}},

    -- light = {intensity = 0.75, size = 8, color = {r = 1.0, g = 1.0, b = 1.0}},
    -- alert_icon_shift = util.by_pixel(-3, -12),
    animation =
    {
      layers =
      {
        {
          filename = "__ColonialCharter__/graphics/entity/colonial-housing/colonial-housing.png",
          width = 98,
          height = 87,
          frame_count = 33,
          line_length = 11,
          animation_speed = 1 / 3,
          shift = util.by_pixel(0, 1.5),
          hr_version =
          {
            filename = "__ColonialCharter__/graphics/entity/colonial-housing/hr-colonial-housing.png",
            width = 194,
            height = 174,
            frame_count = 33,
            line_length = 11,
            animation_speed = 1 / 3,
            shift = util.by_pixel(0, 1.5),
            scale = 0.5
          }
        },   
        {
          filename = "__ColonialCharter__/graphics/entity/colonial-housing/colonial-housing-shadow.png",
          width = 122,
          height = 68,
          frame_count = 1,
          line_length = 1,
          repeat_count = 33,
          animation_speed = 1 / 3,
          shift = util.by_pixel(13, 11),
          draw_as_shadow = true,
          hr_version =
          {
            filename = "__ColonialCharter__/graphics/entity/colonial-housing/hr-colonial-housing-shadow.png",
            width = 242,
            height = 136,
            frame_count = 1,
            line_length = 1,
            repeat_count = 33,
            animation_speed = 1 / 3,
            shift = util.by_pixel(13, 11),
            scale = 0.5,
            draw_as_shadow = true
          }
        }
      }
    },
    crafting_categories = {"colonial-rest"},
    crafting_speed = 5,
    energy_source =
    {
      type = "electric",
      usage_priority = "secondary-input",
      -- emissions = 0.05 / 1.5
    },
    energy_usage = "600kW",
    ingredient_count = 4,
    -- open_sound = { filename = "__base__/sound/machine-open.ogg", volume = 0.85 },
    -- close_sound = { filename = "__base__/sound/machine-close.ogg", volume = 0.75 },
    vehicle_impact_sound =  { filename = "__base__/sound/car-metal-impact.ogg", volume = 0.65 },
    working_sound =
    {
      sound =
      {
        filename = "__base__/sound/lab.ogg",
        volume = 0.7
      },
      apparent_volume = 1
    },
    -- module_specification = { module_slots = 0 }
  },
  
  -- COLONIAL WORKSHOP it's just a copy of housing with diffrent crafting category
  {
    type = "assembling-machine",
    name = "colonial-workshop",
    icon = "__ColonialCharter__/graphics/icons/colonial-workshop.png",
    icon_size = 32,
    flags = {"placeable-neutral", "placeable-player", "player-creation"},
    minable = {hardness = 0.2, mining_time = 0.5, result = "colonial-workshop"},
    max_health = 300,
    corpse = "big-remnants",
    dying_explosion = "medium-explosion",
    collision_box = {{-1.2, -1.2}, {1.2, 1.2}},
    selection_box = {{-1.5, -1.5}, {1.5, 1.5}},
    -- light = {intensity = 0.75, size = 8, color = {r = 1.0, g = 1.0, b = 1.0}},
    -- alert_icon_shift = util.by_pixel(-3, -12),
    animation =
    {
      layers =
      {
        {
          filename = "__ColonialCharter__/graphics/entity/colonial-workshop/colonial-workshop.png",
          width = 98,
          height = 87,
          frame_count = 33,
          line_length = 11,
          animation_speed = 1 / 3,
          shift = util.by_pixel(0, 1.5),
          hr_version =
          {
            filename = "__ColonialCharter__/graphics/entity/colonial-workshop/hr-colonial-workshop.png",
            width = 194,
            height = 174,
            frame_count = 33,
            line_length = 11,
            animation_speed = 1 / 3,
            shift = util.by_pixel(0, 1.5),
            scale = 0.5
          }
        },   
        {
          filename = "__ColonialCharter__/graphics/entity/colonial-workshop/colonial-workshop-shadow.png",
          width = 122,
          height = 68,
          frame_count = 1,
          line_length = 1,
          repeat_count = 33,
          animation_speed = 1 / 3,
          shift = util.by_pixel(13, 11),
          draw_as_shadow = true,
          hr_version =
          {
            filename = "__ColonialCharter__/graphics/entity/colonial-workshop/hr-colonial-workshop-shadow.png",
            width = 242,
            height = 136,
            frame_count = 1,
            line_length = 1,
            repeat_count = 33,
            animation_speed = 1 / 3,
            shift = util.by_pixel(13, 11),
            scale = 0.5,
            draw_as_shadow = true
          }
        }
      }
    },
    crafting_categories = {"colonial-work"},
    crafting_speed = 5,
    energy_source =
    {
      type = "electric",
      usage_priority = "secondary-input",
      -- emissions = 0.05 / 1.5
    },
    energy_usage = "800kW",
    ingredient_count = 4,
    -- open_sound = { filename = "__base__/sound/machine-open.ogg", volume = 0.85 },
    -- close_sound = { filename = "__base__/sound/machine-close.ogg", volume = 0.75 },
    vehicle_impact_sound =  { filename = "__base__/sound/car-metal-impact.ogg", volume = 0.65 },
    working_sound =
    {
      sound =
      {
        filename = "__base__/sound/lab.ogg",
        volume = 0.7
      },
      apparent_volume = 1
    },
    -- module_specification = { module_slots = 0 }
  },

  -- WAITING HALL
  {
    type = "container",
    name = "waiting-hall",
    icon = "__ColonialCharter__/graphics/icons/waiting-hall.png",
    icon_size = 32,
    flags = {"placeable-neutral", "player-creation"},
    minable = {mining_time = 1, result = "waiting-hall"},
    max_health = 200,
    corpse = "medium-remnants",
    open_sound = { filename = "__base__/sound/metallic-chest-open.ogg", volume=0.65 },
    close_sound = { filename = "__base__/sound/metallic-chest-close.ogg", volume = 0.7 },
    resistances =
    {
      {
        type = "fire",
        percent = 80
      },
      {
        type = "impact",
        percent = 30
      }
    },
    collision_box = {{-0.8, -0.8}, {0.8, 0.8}},
    selection_box = {{-1, -1}, {1, 1}},
    -- fast_replaceable_group = "container",
    inventory_size = 32,
    -- item_group_filters = "colonial-charter",
    item_subgroup_filters = {"colonial-colonist"}, -- should restrict it to just colonits, will see if it works. 
    vehicle_impact_sound =  { filename = "__base__/sound/car-metal-impact.ogg", volume = 0.65 },
    picture =
    {
      filename = "__ColonialCharter__/graphics/entity/waiting-hall/waiting-hall.png",
      priority = "extra-high",
      width = 84,
      height = 61,
      shift = {0.245, 0}
      -- shift = util.by_pixel(0, 1.5),
    },
    circuit_wire_connection_point = circuit_connector_definitions["chest"].points,
    circuit_connector_sprites = circuit_connector_definitions["chest"].sprites,
    circuit_wire_max_distance = default_circuit_wire_max_distance
  },

-- LANDING SITE
  {
    type = "assembling-machine",
    name = "landing-site",
    icon = "__ColonialCharter__/graphics/icons/landing-site.png",
    icon_size = 32,
    flags = {"placeable-neutral", "placeable-player", "player-creation"},
    minable = {hardness = 0.2, mining_time = 0.5, result = "landing-site"},
    max_health = 300,
    corpse = "big-remnants",
    dying_explosion = "medium-explosion",
    resistances =
    {
      {
        type = "fire",
        percent = 70
      }
    },
    collision_box = {{-4.2, -4.7}, {4.2, 4.7}},
    selection_box = {{-4.5, -5}, {4.5, 5}},
    alert_icon_shift = util.by_pixel(-3, -12),

    animation =
    {
      layers =
      {
        {
          filename = "__ColonialCharter__/graphics/entity/landing-site/landing-site-night.png",
          width = 352,
          height = 384,
          frame_count = 1,
          repeat_count = 16,
          shift = {0, 0}
          -- frame_count = 18,
          -- line_length = 8,
          -- shift = util.by_pixel(0, 2),
        },
      }
    },
    working_visualisations =
    {
      {
        animation =
        {
          filename = "__ColonialCharter__/graphics/entity/landing-site/smoke-animation.png",
          -- 352,352,16
          line_length = 4,
          width = 352,
          height = 352,
          frame_count = 16,
          animation_speed = 0.8,
          -- animation_speed = 1 / crafting_speed, 
          -- shift = util.by_pixel(0, -15),
        },
      },
    },
    crafting_categories = {"colonial-landing"},
    crafting_speed = 0.5,
    energy_source =
    {
      type = "electric",
      usage_priority = "secondary-input",
      emissions = 0.05 / 1.5
    },
    energy_usage = "90kW",
    ingredient_count = 1,
    open_sound = { filename = "__base__/sound/machine-open.ogg", volume = 0.85 },
    close_sound = { filename = "__base__/sound/machine-close.ogg", volume = 0.75 },
    vehicle_impact_sound =  { filename = "__base__/sound/car-metal-impact.ogg", volume = 0.65 },
    working_sound =
    {
      sound =
      {
        {
          filename = "__base__/sound/assembling-machine-t1-1.ogg",
          volume = 0.8
        },
        {
          filename = "__base__/sound/assembling-machine-t1-2.ogg",
          volume = 0.8
        }
      },
      idle_sound = { filename = "__base__/sound/idle1.ogg", volume = 0.6 },
      apparent_volume = 1.5
    }
  },

  -- COMS DISH
  {
    type = "assembling-machine",
    name = "coms-dish",
    icon = "__ColonialCharter__/graphics/icons/coms-dish.png",
    icon_size = 32,
    flags = {"placeable-neutral", "placeable-player", "player-creation"},
    minable = {hardness = 0.2, mining_time = 0.5, result = "coms-dish"},
    max_health = 300,
    corpse = "big-remnants",
    dying_explosion = "medium-explosion",
    collision_box = {{-1.2, -1.2}, {1.2, 1.2}},
    selection_box = {{-1.5, -1.5}, {1.5, 1.5}},
    -- light = {intensity = 0.75, size = 8, color = {r = 1.0, g = 1.0, b = 1.0}},
    -- alert_icon_shift = util.by_pixel(-3, -12),
    animation =
    {
      layers =
      {
        {
          filename = "__ColonialCharter__/graphics/entity/coms-dish/coms-dish.png",
          width = 171,
          height = 190,
          frame_count = 5,
          line_length = 5,
          animation_speed = 1 / 90,
                              -- 135 38 - I have no idea how this shift thing works, or what the units are
          shift = util.by_pixel(18, -47), 
          run_mode = "forward-then-backward",
          hr_version =
          {
            filename = "__ColonialCharter__/graphics/entity/coms-dish/hr-coms-dish.png",
            width = 343,
            height = 381,
            frame_count = 5,
            line_length = 5,
            animation_speed = 1 / 90,
            shift = util.by_pixel(18, -47),
            run_mode = "forward-then-backward",
            scale = 0.5
          }
        },   
        {
          filename = "__ColonialCharter__/graphics/entity/coms-dish/coms-dish-shadow.png",
          width = 289,
          height = 122,
          frame_count = 5,
          line_length = 5,
          animation_speed = 1 / 90,
                                --214 53
          shift = util.by_pixel(110, 19), 
          run_mode = "forward-then-backward",
          draw_as_shadow = true,
          hr_version =
          {
            filename = "__ColonialCharter__/graphics/entity/coms-dish/hr-coms-dish-shadow.png",
            width = 578,
            height = 244,
            frame_count = 5,
            line_length = 5,
            animation_speed = 1 / 90,
            shift = util.by_pixel(110, 19), 
            run_mode = "forward-then-backward",
            scale = 0.5,
            draw_as_shadow = true
          }
        }
      }
    },
    crafting_categories = {"colonial-coms"},
    crafting_speed = 5,
    energy_source =
    {
      type = "electric",
      usage_priority = "secondary-input",
      -- emissions = 0.05 / 1.5
    },
    energy_usage = "2800kW",
    ingredient_count = 1,
    -- open_sound = { filename = "__base__/sound/machine-open.ogg", volume = 0.85 },
    -- close_sound = { filename = "__base__/sound/machine-close.ogg", volume = 0.75 },
    vehicle_impact_sound =  { filename = "__base__/sound/car-metal-impact.ogg", volume = 0.65 },
    working_sound =
    {
      sound =
      {
        filename = "__base__/sound/lab.ogg",
        volume = 0.7
      },
      apparent_volume = 1
    },
    -- module_specification = { module_slots = 0 }
    },

  -- HYDROPONICS BUILDING
  -- TODO: add pipe connections they should be located on the left and right edges 2.5 squars down from the center. 
    {
      type = "assembling-machine",
      name = "hydroponics-building",
      icon = "__ColonialCharter__/graphics/icons/hydroponics-building.png",
      icon_size = 32,
      flags = {"placeable-neutral", "placeable-player", "player-creation"},
      minable = {hardness = 0.2, mining_time = 0.5, result = "hydroponics-building"},
      max_health = 300,
      corpse = "big-remnants",
      dying_explosion = "medium-explosion",
      collision_box = {{-4.2, -3.2}, {4.2, 3.2}},
      selection_box = {{-4.5, -3.5}, {4.5, 3.5}},
      -- light = {intensity = 0.75, size = 8, color = {r = 1.0, g = 1.0, b = 1.0}},
      -- alert_icon_shift = util.by_pixel(-3, -12),
      animation =
      {
        layers =
        {
          {
            filename = "__ColonialCharter__/graphics/entity/hydroponics-building/growing.png",
            width = 258,
            height = 186.5,
            frame_count = 16,
            line_length = 4,
            animation_speed = 1/30,
            shift = util.by_pixel(-1.5, -0.25)
          },
          {
            filename = "__ColonialCharter__/graphics/entity/hydroponics-building/glass-dome.png",
            width = 288,
            height = 224,
            frame_count = 1,
            repeat_count = 16,
            -- frame_count = 16,
            -- line_length = 4,
            -- animation_speed = 1 / 30,
            hr_version =
            {
              filename = "__ColonialCharter__/graphics/entity/hydroponics-building/hr-glass-dome.png",
              width = 576,
              height = 448,
              frame_count = 1,
              repeat_count = 16,
              -- frame_count = 16,
              -- line_length = 4,
              -- animation_speed = 1 / 30,
              scale = 0.5
            }
          },   
          {
            filename = "__ColonialCharter__/graphics/entity/hydroponics-building/hydroponics-building-shadow.png",
            width = 320,
            height = 207,
            frame_count = 1,
            repeat_count = 16,
            shift = util.by_pixel(16.75, 5.5), 
            draw_as_shadow = true,
            hr_version =
            {
              filename = "__ColonialCharter__/graphics/entity/hydroponics-building/hr-hydroponics-building.png",
              width = 641,
              height = 414,
              frame_count = 1,
              -- repeat_count = 16,
              shift = util.by_pixel(16.75, 5.5), 
              scale = 0.5,
              draw_as_shadow = true
            }
          }
        }
      },
      crafting_categories = {"colonial-agriculture"},
      crafting_speed = 5,
      energy_source =
      {
        type = "electric",
        usage_priority = "secondary-input",
        -- emissions = 0.05 / 1.5
      },
      energy_usage = "800kW",
      ingredient_count = 5,
      -- open_sound = { filename = "__base__/sound/machine-open.ogg", volume = 0.85 },
      -- close_sound = { filename = "__base__/sound/machine-close.ogg", volume = 0.75 },
      vehicle_impact_sound =  { filename = "__base__/sound/car-metal-impact.ogg", volume = 0.65 },
      working_sound =
      {
        sound =
        {
          filename = "__base__/sound/lab.ogg",
          volume = 0.7
        },
        apparent_volume = 1
      },
      -- module_specification = { module_slots = 0 }
    },

})
